### ISSUE TITLE
stack not being called correctly

### ISSUE THREAD

**Issue by ruslantalpa** [mvoidex/hsdev#38]
i think this line is wrong
https://github.com/mvoidex/hsdev/blob/master/src/HsDev/Stack.hs#L67
liftIO $ readProcess stackExe (cmd' ++ ["--compiler", comp, "--arch", stackArch]) ""

stack wants it's parameters before the main command
for example this works
`/bin/stack --stack-yaml ./stack.yaml --compiler "ghc-7.10.3" --arch x86_64 path`

this doesn't 
`/bin/stack path --stack-yaml ./stack.yaml --compiler "ghc-7.10.3" --arch x86_64` and fails with

and fails with
```
Invalid argument `ghc-7.10.3'
Auxiliary command not found in path `stack-path'
File does not exist or is not a regular file `path'
```

this manifests itself here
```
hsdev scan --path=./
{"notify":[{"status":"working","progress":null,"name":"/Users/username/projectdir","task":"scanning","type":"path"}]}
{"notify":[{"status":{"error":"io error","msg":"readCreateProcess: /bin/stack \"path\" \"--stack-yaml\" \"/Users/username/projectdir/stack.yaml\" \"--compiler\" \"ghc-7.10.3\" \"--arch\" \"x86_64\" (exit 1): failed"},"progress":null,"name":"/Users/username/projectdir","task":"scanning","type":"path"}]}
[]
```

Am i missing something and this actually works for someone?
I am trying to get this working with sublime.
Do you want a PR for this (though it's trivial) and once it's merged, any pointers on how to have this new version working/installed (i am spoiled and only used stack when working with haskell)?
Thank you.

PS: also, before this, hsdev did not find stack in /usr/local/bin/stack but after i added a symlink to /usr/bin/stack it worked

**Comment by mvoidex** [comment]
What's your stack version (`stack --version`)?
My is `Version 1.2.0 x86_64 hpack-0.14.0` and both commands works.

**Comment by ruslantalpa** [comment]
Version 1.1.0 x86_64 hpack-0.13.0

hmm :) 
maybe an upgrade will work (i was already half way converting this to stack and trying to compile)
:)

**Comment by ruslantalpa** [comment]
just out of curiosity i'll leave it and see if it builds and also change the order to see if it will work with the old stack.
if it ends up working, do you want a PR with a stack.yaml and this modification to support older stack versions?

**Comment by mvoidex** [comment]
Ok, PR will be good.
As long as it is not critical issue, I won't bump hsdev's version right now, if new stack version will work for you. There are some stack-related issues I want to fix before (despite the fact i don't have much time now).

**Comment by ruslantalpa** [comment]
It compiled and stack worked.
Any idea why i would get
```
remote: Your unique error code: fb9030d713453a500e2d62315535a497
To https://github.com/ruslantalpa/hsdev.git
 ! [remote rejected] master -> master (pre-receive hook declined)
```
when pushing to my own fork?

**Comment by mvoidex** [comment]
Hm, no idea

**Comment by ruslantalpa** [comment]
check your settings/branches, anything there saying "protected branch"?

**Comment by mvoidex** [comment]
No, I haven't altered any settings

> No protected branches yet.

**Comment by ruslantalpa** [comment]
i'll try to investigate and if all fails, i'll paste a patch here

**Comment by bscottm** [comment]
Actually, this seems to happen with the latest stack:

    Version 1.2.0, Git revision 241cd07d576d9c0c0e712e83d947e3dd64541c42 (4054 commits) x86_64 hpack-0.14.0

**Comment by bscottm** [comment]
Doing a little more diagnosis with SublimeHaskell, SublimeHaskell doesn't consistently invoke `stack` to get all of the paths:

    /<path-to-stack>/stack "--compiler" "ghc-8.0.1" "--arch" "x86_64" "path"

Consequently, `hsdev` tries to invoke `ghc`, which naturally fails. There also seems to be some issue with `HsDev.Stack.stack`, which I haven't been able to trace down yet. I'll have to trace a bit using `readCreateProcess` to capture `stderr` to see what the problem is, once I have a better idea with what's going on in SublimeHaskell.

**Comment by bscottm** [comment]
@mvoidex, @ruslantalpa: My educated guess is that there is a sequencing issue in `hsdev`. Looking at the logs, `hsdev` runs several `ghc` command sequences. In my case, `ghc` is installed in the `stack` sandbox, which is not under `${PWD}/.stack-work`. Consequently, invoking `ghc` will always fail, hence, the IO exceptions that mention `hPutChar`. (`hsdev` should fail more gracefully if there it can't find `ghc`, but that's a separate issue.)

Basically, `hsdev` needs to determine `ghc`'s location first before doing anything else. It's something I'll look into in the next few days as time permits.

Also, it might be a good idea to capture the entire association list from `stack path`, because one of the paths is `compiler-exe`, if `ghc` cannot be found on `PATH`.

**Comment by bscottm** [comment]
@mvoidex: Should I open a separate ticket for this issue?

**Comment by mvoidex** [comment]
@bscottm, yes, separate issue will be good. Sorry for long silence.

**Comment by mvoidex** [comment]
Seems, that this issue can be closed?
